bpf: fix gcc-8 warnings

nirmoy opened the pull request:
gcc-8 throws following warning without this changes:
cilium-map-migrate.c: In function ‘bpf_handle_finalize’:
cilium-map-migrate.c:373:37: warning: ‘:’ directive output may be truncated writing 1 byte into a region of size between 0 and 4096 [-Wformat-truncation=]
  snprintf(file, sizeof(file), "%s/%s:%s", fs_base, name,
                                     ^
cilium-map-migrate.c:373:2: note: ‘snprintf’ output 10 or more bytes (assuming 4106) into a destination of size 4097
  snprintf(file, sizeof(file), "%s/%s:%s", fs_base, name,
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    STATE_PENDING);
    ~~~~~~~~~~~~~~
cilium-map-migrate.c:384:38: warning: ‘snprintf’ output may be truncated before the last format character [-Wformat-truncation=]
   snprintf(dest, sizeof(dest), "%s/%s", fs_base, name);
                                      ^
cilium-map-migrate.c:384:3: note: ‘snprintf’ output 2 or more bytes (assuming 4098) into a destination of size 4097
   snprintf(dest, sizeof(dest), "%s/%s", fs_base, name);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cilium-map-migrate.c: In function ‘bpf_handle_pending’:
cilium-map-migrate.c:332:37: warning: ‘snprintf’ output may be truncated before the last format character [-Wformat-truncation=]
  snprintf(file, sizeof(file), "%s/%s", fs_base, name);
                                     ^
cilium-map-migrate.c:332:2: note: ‘snprintf’ output 2 or more bytes (assuming 4098) into a destination of size 4097
  snprintf(file, sizeof(file), "%s/%s", fs_base, name);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cilium-map-migrate.c:358:35: warning: ‘%s’ directive output may be truncated writing 7 bytes into a region of size between 0 and 4096 [-Wformat-truncation=]
  snprintf(dest, sizeof(dest), "%s:%s", file, STATE_PENDING);
                                   ^~
cilium-map-migrate.c:358:2: note: ‘snprintf’ output between 9 and 4105 bytes into a destination of size 4097
  snprintf(dest, sizeof(dest), "%s:%s", file, STATE_PENDING);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cilium-map-migrate.c: In function ‘main’:
cilium-map-migrate.c:489:9: warning: ‘fn’ may be used uninitialized in this function [-Wmaybe-uninitialized]
   ret = cb(ctx, map, name, exit);
         ^~~~~~~~~~~~~~~~~~~~~~~~
cilium-map-migrate.c:555:21: note: ‘fn’ was declared here
  bpf_handle_state_t fn;

Signed-off-by: Nirmoy Das <ndas@suse.de>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4785)
<!-- Reviewable:end -->

borkmann commented on bpf/cilium-map-migrate.c:
```diff
@@ -307,17 +307,18 @@ typedef int (*bpf_handle_state_t)(struct bpf_elf_ctx *ctx,
-		strcpy(fs_base, "/sys/fs/bpf/tc/globals");
+		strcpy(file, "/sys/fs/bpf/tc/globals");
 	}
+	fs_base = strdup(file);
```
Since this is just a synthetical case, could we just use `-Wnoformat-truncation` and let subsequent open request fail? After all the workaround here is not really fixing the warning either since you'd need 2*PATH_MAX +1 buffer which couldn't even be handled by kernel's path lookup anyway.

nirmoy replied:
all the file name variable are in PATH_MAX + 1 so subsequent open/stat/rename will fail.

borkmann replied:
Right, which is fine, but the "only" thing this bit above does is to hide the fact from gcc that fs_base would be of size `PATH_MAX + 1`, since we now need to allocate it dynamically, hence, gcc cannot statically verify, and thus warn about it. Meaning, this bit is for the purpose of silencing the gcc warning in which case it would be much more straight forward to just add to CFLAGS disabling of this warning.

It also introduces two actual bugs: i) strdup() can fail but is not checked here (unlikely but not impossible), ii) we should also free(fs_base) in case fn == NULL test below.

nirmoy replied:
Thanks  @borkmann I updated the pr with Wno-format-truncation tested with gcc-7/8

borkmann commented on bpf/cilium-map-migrate.c:
```diff
@@ -552,7 +553,7 @@ static int migrate_state(const char *pathname, bpf_handle_state_t cb, int exit)
 {
 	const char *pathname = NULL;
-	bpf_handle_state_t fn;
+	bpf_handle_state_t fn = NULL;
```
This is not a complete fix. We should bail out if fn is NULL before we go into migrate_state() similar to default switch case above.

nirmoy replied:
updated the PR with NULL check, Thanks

nirmoy committed:
bpf: fix gcc-8 warnings

gcc-8 throws following warning without this changes:
cilium-map-migrate.c: In function ‘bpf_handle_finalize’:
cilium-map-migrate.c:373:37: warning: ‘:’ directive output may be truncated writing 1 byte into a region of size between 0 and 4096 [-Wformat-truncation=]
  snprintf(file, sizeof(file), "%s/%s:%s", fs_base, name,
                                     ^
cilium-map-migrate.c:373:2: note: ‘snprintf’ output 10 or more bytes (assuming 4106) into a destination of size 4097
  snprintf(file, sizeof(file), "%s/%s:%s", fs_base, name,
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    STATE_PENDING);
    ~~~~~~~~~~~~~~
cilium-map-migrate.c:384:38: warning: ‘snprintf’ output may be truncated before the last format character [-Wformat-truncation=]
   snprintf(dest, sizeof(dest), "%s/%s", fs_base, name);
                                      ^
cilium-map-migrate.c:384:3: note: ‘snprintf’ output 2 or more bytes (assuming 4098) into a destination of size 4097
   snprintf(dest, sizeof(dest), "%s/%s", fs_base, name);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cilium-map-migrate.c: In function ‘bpf_handle_pending’:
cilium-map-migrate.c:332:37: warning: ‘snprintf’ output may be truncated before the last format character [-Wformat-truncation=]
  snprintf(file, sizeof(file), "%s/%s", fs_base, name);
                                     ^
cilium-map-migrate.c:332:2: note: ‘snprintf’ output 2 or more bytes (assuming 4098) into a destination of size 4097
  snprintf(file, sizeof(file), "%s/%s", fs_base, name);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cilium-map-migrate.c:358:35: warning: ‘%s’ directive output may be truncated writing 7 bytes into a region of size between 0 and 4096 [-Wformat-truncation=]
  snprintf(dest, sizeof(dest), "%s:%s", file, STATE_PENDING);
                                   ^~
cilium-map-migrate.c:358:2: note: ‘snprintf’ output between 9 and 4105 bytes into a destination of size 4097
  snprintf(dest, sizeof(dest), "%s:%s", file, STATE_PENDING);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cilium-map-migrate.c: In function ‘main’:
cilium-map-migrate.c:489:9: warning: ‘fn’ may be used uninitialized in this function [-Wmaybe-uninitialized]
   ret = cb(ctx, map, name, exit);
         ^~~~~~~~~~~~~~~~~~~~~~~~
cilium-map-migrate.c:555:21: note: ‘fn’ was declared here
  bpf_handle_state_t fn;

Signed-off-by: Nirmoy Das <ndas@suse.de>

borkmann reviewed (approved):
LGTM, thanks!
